geps/gep-1713.md

```diff
@@ -0,0 +1,103 @@
+# GEP-1713: Standard Mechanism to Merge Multiple Gateways
+
+* Issue: [#1713/kubernetes-sigs/gateway-api/issues/1713)
+* Status: Provisional
+
+(See status definitions [here](overview.md#status).)
+
+## TLDR
+
+The Gateway Resource is a contention point since it is the only place to attach listeners with certificates. We propose a mechanism to allow distinct Gateway resources to be logically merged.
+
+## Goals
+
+- Define a mechanic to merge multiple Gateways (logically)
+- Define a set of acceptable properties that can be merged and their semantics
+
+## Non-Goals
+
+- Apply a Gateway resource onto N distinct gateways (one to many)
+
+## Introduction
+
+Knative generates on demand per-service certificates using HTTP-01 challenges. There can be O(1000) Knative Services in the cluster which means we have O(1000) distinct certificates. Thus updating a single Gateway resource with this many certificates is a contention point and inhibits horizontal scaling of our controllers.
+
+The spec currently has language to indicate implemenations `MAY` merge Gateways resources but the mechanic isn't defined. 
+https://github.com/kubernetes-sigs/gateway-api/blob/541e9fc2b3c2f62915cb58dc0ee5e43e4096b3e2/apis/v1beta1/gateway_types.go#L76-L78
+
+## Future Requirement Level
+
+We want to keep this API very simple so that the merging requirement level could increase from `MAY` to `MUST`
+
+## API
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: parent-gateway
```

Comment:
can the parent gateway choose which children gateways it wants to be linked with ?
imo this can be mitigated by an defining the intent to merge at a top level (or even sibling level) rather than going bottom up (child to parent)

Reply:
1. have a `mergeable`/`combinable` field at the `GatewayClass` level that need to be set to `true` or some enum (which defines merge capability)
or
2. add a `group` / `scope` field at the `Gateway` level as suggested by you earlier. But @youngnick raised a good point earlier [here](https://github.com/kubernetes-sigs/gateway-api/pull/1653#issuecomment-1447523825) that this style makes grouping hard to comprehend.

Both the approaches would require a new status condition at the `Gateway` level to be created to signal outcome of the merge.

Reply:
I'm considering the split between gateways to just be 'logical' - meaning how do I take a single gateway and describe it using N objects. In contrast to Jon's proposal about how N distinct gateways can merge into the same infrastructure (https://github.com/kubernetes-sigs/gateway-api/pull/1757)

Based on that regarding 1. - I would want to avoid defining things at the GatewayClass. I think that answers 2 as well since you wouldn't merge Gateways with different addresses/scopes.

> Both the approaches would require a new status condition at the Gateway level to be created to signal outcome of the merge.

So I added the following clause to help simplify this 
> Each `Gateway` will only list listeners in their status if there is a corresponding entry in their spec.

Do you envision other status conditions?

Reply:
> can the parent gateway choose which children gateways it wants to be linked with ?

If this is still needed, I'd imagine it could look similar to [`AllowedRoutes`](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.AllowedRoutes), adding an optional field to `Gateway.spec` named something like `MergeableGateways` with a `namespaces` field similar to [`RouteNamespaces`](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.RouteNamespaces).

The biggest difference is that the default value for `MergeableGateways` (to ensure allowing merge semantics remains opt-in, and to raise a status condition if configured in spec but not supported by the implementation) should probably be:
```
from:
  namespaces: "None"
```
which would necessitate the addition of an additional value for [`FromNamespaces`](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.FromNamespaces)